homotopy-graphics/src/style.rs

```diff
@@ -1,20 +1,66 @@
+use std::{fmt, str::FromStr};
+
 use homotopy_core::Generator;
+use serde::{Deserialize, Serialize};
 
 pub trait GeneratorStyle {
     fn label(&self) -> Option<String>;
     fn shape(&self) -> Option<VertexShape>;
-    // TODO(thud): migrate color-related code to here for consistency
+    fn color(&self) -> Color;
 }
 
 pub trait SignatureStyleData {
-    type T: GeneratorStyle;
+    type Style: GeneratorStyle;
+
+    fn generator_style(&self, g: Generator) -> Option<&Self::Style>;
 
-    fn generator_style(&self, g: Generator) -> Option<&Self::T>;
+    // It would be nice if the following could be an iterator but the generics get complex fast
+    fn generators(&self) -> Vec<Generator>;
```

Comment:
If you look at where we use this, it's always the same pattern:

```rust
for g in signature_styles.generators() {
    let style = signature_styles.generator_style(g);
    ...
}
```

But `generator_style` expands to `generator_info` which works by iterating over the generators again until it finds the right one. So the whole thing is quadratic time.

We can make this better if we make this function return a list of `Style` rather than `Generator`.

Reply:
And probably change the name to `generator_styles` or something.

Reply:
Sorry it should return pairs `(Generator, Style)` as we need the generator as well.

Reply:
Yup, good point